### PR TITLE
Small additions to QgsBox3d and QgsOrientedBox3D API

### PR DESCRIPTION
**Pull request by wonder-sk**
A couple of utility functions that will be useful when dealing with 3D Tiles:

- operators to move QgsBox3d by a QgsVector3D
- reproject QgsOrientedBox3D and get its axis-aligned bounding box
- get length of sides of QgsOrientedBox3D
- transform QgsOrientedBox3D with a QgsMatrix4x4

### CODE REVIEW

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 1]
```diff
@@ -93,3 +95,56 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+  QgsVector3D axis1( mHalfAxes[0], mHalfAxes[1], mHalfAxes[2] );
+  QgsVector3D axis2( mHalfAxes[3], mHalfAxes[4], mHalfAxes[5] );
+  QgsVector3D axis3( mHalfAxes[6], mHalfAxes[7], mHalfAxes[8] );
+  return QgsVector3D( 2 * axis1.length(), 2 * axis2.length(), 2 * axis3.length() );
+}
+
+
+QgsBox3d QgsOrientedBox3D::reprojectedExtent( const QgsCoordinateTransform &ct ) const
+{
+  // reproject corners from ECEF to planar CRS
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.cpp:111]:
```suggestion
  // reproject corners to destination CRS
```

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 2]
```diff
@@ -93,3 +95,56 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+  QVector<QgsVector3D> c = corners();
+  for ( int i = 0; i < c.count(); ++i )
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.cpp:113]:
```suggestion
  Q_ASSERT( c.count() == 8 );
  for ( int i = 0; i < 8; ++i )
```

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 3]
```diff
@@ -93,3 +95,56 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+    if ( v.y() > v1.y() ) v1.setY( v.y() );
+    if ( v.z() > v1.z() ) v1.setZ( v.z() );
+  }
+  return QgsBox3d( v0.x(), v0.y(), v0.z(), v1.x(), v1.y(), v1.z() );
+}
+
+QgsOrientedBox3D QgsOrientedBox3D::transformed( const QgsMatrix4x4 &tr ) const
+{
+  const double *ptr = tr.constData();
+  QgsMatrix4x4 mm( ptr[0], ptr[4], ptr[8], 0,
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.cpp:135]:
```suggestion
  const QgsMatrix4x4 mm( ptr[0], ptr[4], ptr[8], 0,
```

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 4]
```diff
@@ -93,3 +95,56 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+                   ptr[1], ptr[5], ptr[9], 0,
+                   ptr[2], ptr[6], ptr[10], 0,
+                   0, 0, 0, 1 );
+
+  QgsVector3D trCenter = tr.map( QgsVector3D( mCenter[0], mCenter[1], mCenter[2] ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.cpp:140]:
```suggestion
  const QgsVector3D trCenter = tr.map( QgsVector3D( mCenter[0], mCenter[1], mCenter[2] ) );
```

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 5]
```diff
@@ -93,3 +95,56 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+
+  QgsVector3D col1 = mm.map( QgsVector3D( mHalfAxes[0], mHalfAxes[1], mHalfAxes[2] ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.cpp:142]:
```suggestion
  const QgsVector3D col1 = mm.map( QgsVector3D( mHalfAxes[0], mHalfAxes[1], mHalfAxes[2] ) );
```

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 6]
```diff
@@ -93,3 +95,56 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+  QgsVector3D col2 = mm.map( QgsVector3D( mHalfAxes[3], mHalfAxes[4], mHalfAxes[5] ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.cpp:143]:
```suggestion
  const QgsVector3D col2 = mm.map( QgsVector3D( mHalfAxes[3], mHalfAxes[4], mHalfAxes[5] ) );
```

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 7]
```diff
@@ -93,3 +95,56 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+  QgsVector3D col3 = mm.map( QgsVector3D( mHalfAxes[6], mHalfAxes[7], mHalfAxes[8] ) );
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.cpp:144]:
```suggestion
  const QgsVector3D col3 = mm.map( QgsVector3D( mHalfAxes[6], mHalfAxes[7], mHalfAxes[8] ) );
```

File: src/core/geometry/qgsorientedbox3d.h [hunk 8]
```diff
@@ -122,6 +124,22 @@ class CORE_EXPORT QgsOrientedBox3D
...
+    /**
+     * Returns size of sides of the box.
+     */
+    QgsVector3D size() const;
+
+    /**
+     * Reprojects corners of this box using the given coordinate transform
+     * and returns axis-aligned box containing reprojected corners.
+     */
+    QgsBox3d reprojectedExtent( const QgsCoordinateTransform &ct ) const;
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.h:136]:
```suggestion
    /**
     * Reprojects corners of this box using the given coordinate \a transform
     * and returns axis-aligned box containing reprojected corners.
     * \throws QgsCsException
     */
    QgsBox3d reprojectedExtent( const QgsCoordinateTransform &transform ) const SIP_THROW( QgsCsException );
```

File: src/core/geometry/qgsorientedbox3d.h [hunk 9]
```diff
@@ -122,6 +124,22 @@ class CORE_EXPORT QgsOrientedBox3D
...
+
+    /**
+     * Returns box transformed by a 4x4 matrix.
+     */
+    QgsOrientedBox3D transformed( const QgsMatrix4x4 &tr ) const;
```
Comment (nyalldawson) [inline src/core/geometry/qgsorientedbox3d.h:141]:
```suggestion
    QgsOrientedBox3D transformed( const QgsMatrix4x4 &transform ) const;
```

**Review by nyalldawson — APPROVED** [review]

File: src/core/geometry/qgsorientedbox3d.cpp [hunk 10]
```diff
@@ -93,3 +95,57 @@ QVector<QgsVector3D> QgsOrientedBox3D::corners() const
...
+  for ( int i = 0; i < 8; ++i )
+  {
+    c[i] = ct.transform( c[i] );
+  }
+
+  // find AABB for the 8 transformed points
+  QgsVector3D v0 = c[0], v1 = c[0];
+  for ( const QgsVector3D &v : c )
+  {
+    if ( v.x() < v0.x() ) v0.setX( v.x() );
```
Comment (benoitdm-oslandia) [inline src/core/geometry/qgsorientedbox3d.cpp:123]: Replace by std::min/std::max?

```suggestion
    v0.setX( std::min( v0.x(), v.x() ) );
```

**Review by benoitdm-oslandia — COMMENTED** [review]
LGTM

File: src/core/geometry/qgsorientedbox3d.cpp (hunk 10 above)
Comment (wonder-sk) [reply src/core/geometry/qgsorientedbox3d.cpp:123]: That would work as well, but is there any advantage over the current implementation?

Comment (benoitdm-oslandia) [reply src/core/geometry/qgsorientedbox3d.cpp:123]: readability. But a bit too late

**Pull request merged**